ballerina/http_commons.bal

```diff
@@ -56,16 +58,48 @@ isolated function buildRequest(RequestMessage message) returns Request|ClientErr
     } else if (message is mime:Entity[]) {
         request.setBodyParts(message);
     } else {
-        var result = trap val:toJson(message);
-        if (result is error) {
-            return error InitializingOutboundRequestError("json conversion error: " + result.message(), result);
-        } else {
-            request.setJsonPayload(result);
+        match mediaType {
+            mime:APPLICATION_FORM_URLENCODED => {
+                string payload = check processUrlEncodedContent(message);
+                request.setTextPayload(payload, mime:APPLICATION_FORM_URLENCODED);
+            }
+            _ => {
+                json payload = check processJsonContent(message);
+                request.setJsonPayload(payload);
+            }
         }
     }
     return request;
 }
 
+isolated function processUrlEncodedContent(anydata message) returns string|ClientError {
+    if message is map<string> {
+        string[] messageParams = message.entries().toArray().'map(constructFormEntry);
+        string payload = strings:'join("&", ...messageParams);
+        string|error encodedContent = url:encode(payload, "UTF-8");
+        if encodedContent is string {
+            return encodedContent;
+        } else {
+            return error InitializingOutboundRequestError("content encoding error: " + encodedContent.message(), encodedContent);
+        }
+    } else {
+        return error InitializingOutboundRequestError("unsupported content for application/x-www-form-urlencoded media type");
+    }
+}
+
+isolated function constructFormEntry([string, string] keyValPair) returns string {
+    return string`${keyValPair[0]}=${keyValPair[1]}`;
+}
+
+isolated function processJsonContent(anydata message) returns json|ClientError {
+    var result = trap val:toJson(message);
+    if (result is error) {
+        return error InitializingOutboundRequestError("json conversion error: " + result.message(), result);
+    } else {
+        return result;
+    }
```

Comment:
I think now we can simple return without an `else`

Reply:
Fixed with [dbce5f84fed5da34d8e585ca4c4a2631a14a00e2](https://github.com/ballerina-platform/module-ballerina-http/pull/797/commits/dbce5f84fed5da34d8e585ca4c4a2631a14a00e2)